java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -35,41 +34,79 @@ import kotlinx.coroutines.withContext
  */
 class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     val storageKey: StorageKey,
-    val backingType: Type,
-    val callbackFactory: (String) -> ProxyCallback<Data, Op, T>
+    val backingType: Type
 ) {
     private val storeMutex = Mutex()
     private val log = TaggedLog { "DirectStoreMuxer" }
+    private val proxyManager = RandomProxyCallbackManager<Data, Op, T>(
+        "backing",
+        Random
+    )
 
     // TODO(b/158262634): Make this CacheMap Weak.
     /* internal */ val stores = LruCacheMap<String, StoreRecord<Data, Op, T>>(
         50,
         livenessPredicate = { _, sr -> !sr.store.closed }
-    ) { _, sr -> closeStore(sr) }
+    ) { muxId, sr -> closeStore(muxId, sr) }
+
+    // Keeps track of all the stores a callbackId is registered to. Used when removing an
+    // observer (off method), as that requires unregistering to each store it is registered with.
+    private val callbackIdToMuxIdMap = mutableMapOf<Int, MutableSet<String>>()
+
+    fun on(callback: ProxyCallback<Data, Op, T>, callbackToken: Int? = null): Int {
+        // Direct Store Muxer must generate the callback token to ensure no duplicates.
+        check(callbackToken == null) { "token must be generated by Direct Store Muxer" }
+        synchronized(proxyManager) {
+            val callbackId = proxyManager.register(callback, callbackToken)
+            callbackIdToMuxIdMap[callbackId] = mutableSetOf()
+            return callbackId
+        }
+    }
+
+    fun off(callbackToken: Int) {
+        synchronized(proxyManager) {
+            for (muxId in callbackIdToMuxIdMap.getOrElse(callbackToken, { mutableSetOf() })) {
+                val (idSet, store) = checkNotNull(stores[muxId]) { "store not found" }
+                store.off(callbackToken)
+                idSet.remove(callbackToken)
+            }
+            callbackIdToMuxIdMap.remove(callbackToken)
+            proxyManager.unregister(callbackToken)
+        }
+    }
 
     /** Safely closes a [DirectStore] and cleans up its resources. */
-    private fun closeStore(storeRecord: StoreRecord<*, *, *>) {
+    private fun closeStore(muxId: String, storeRecord: StoreRecord<*, *, *>) {
         if (!storeRecord.store.closed) {
-            log.debug { "close the store(${storeRecord.id})" }
+            log.debug { "close the store($muxId)" }
 
             try {
+                for (id in storeRecord.idSet) {
+                    storeRecord.store.off(id)
+                    callbackIdToMuxIdMap[id]?.remove(muxId)
```

Comment:
Here's another mutation of the callbackIdToMuxIdMap outside of a lock.

Reply:
I've put it inside a synchronized block 👍